<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>lib/errors.js - Documentation</title>

    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="https://code.ionicframework.com/ionicons/2.0.1/css/ionicons.min.css">
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="Habitica.html">Habitica</a><ul class='methods'><li data-type='method'><a href="Habitica.html#del">del</a></li><li data-type='method'><a href="Habitica.html#get">get</a></li><li data-type='method'><a href="Habitica.html#getOptions">getOptions</a></li><li data-type='method'><a href="Habitica.html#localLogin">localLogin</a></li><li data-type='method'><a href="Habitica.html#post">post</a></li><li data-type='method'><a href="Habitica.html#put">put</a></li><li data-type='method'><a href="Habitica.html#register">register</a></li><li data-type='method'><a href="Habitica.html#setOptions">setOptions</a></li></ul></li><li><a href="HabiticaApiError.html">HabiticaApiError</a></li><li><a href="UnknownConnectionError.html">UnknownConnectionError</a></li></ul>
</nav>

<div id="main">
    
    <h1 class="page-title">lib/errors.js</h1>
    

    



    
    <section>
        <article>
            <pre class="prettyprint source linenums"><code>'use strict'

function CustomError (message) {
  /** A translated error message you can provide for your user
   * @memberof HabiticaApiError
   */
  this.message = message

  Error.captureStackTrace &amp;&amp; Error.captureStackTrace(this, this.constructor)
}

CustomError.prototype = Object.create(Error.prototype)

/**
 * @constructor HabiticaApiError
 * @description Returned when the API request returns a HTTP status between 400 and 500
 *
 * @example &lt;caption>Use the error objects on the Habitica package to write an error handler&lt;/caption>
 * var Habitica = require('habitica')
 * var api = new Habitica({
 *   // setup client
 * })
 *
 * api.get('/user').then(() => {
 *   // will never get here
 * }).catch((err) => {
 *   if (err instanceof Habitica.ApiError) {
 *     // likely a validation error from
 *     // the API request
 *     console.log(err.message)
 *   } else if (err instanceof Habitica.UnknownConnectionError) {
 *     // either the Habitica API is down
 *     // or there is no internet connection
 *     console.log(err.originalError)
 *   } else {
 *     // there is something wrong with your integration
 *     // such as a syntax error or other problem
 *     console.log(err)
 *   }
 * })
 */
function HabiticaApiError (options) {
  options = options || {}

  CustomError.call(this, options.message)

  /** The status code of the HTTP request. Will be a number >= `400` and &lt;= `500`
   * @memberof HabiticaApiError
   */
  this.status = options.status

  /** A type coresponding to the status code. For instance, an error with a status of `404` will be type `NotFound`
   * @memberof HabiticaApiError
   */
  this.type = options.type
  this.name = 'HabiticaApi' + options.type + 'Error'
}

HabiticaApiError.prototype = Object.create(CustomError.prototype)

/**
 * @constructor UnknownConnectionError
 *
 * @description Returned when an error could not be parsed from a failed request. Most likely when there is not an internet connection.
 * @example &lt;caption>See {@link HabiticaApiError} for a full example of how to use it when handling request errors.&lt;/caption>
 * api.get('/user').then(() => {
 *   // assuming there is no internet
 *   // connection, so user never
 *   // gets here
 * }).catch((err) => {
 *   err.message // 'An unknown error occurred'
 *   console.error(err.originalError)
 * })
 */
function UnknownConnectionError (err) {
  CustomError.call(this, 'An unknown error occurred')

  /** The original error from the failed request
   * @memberof UnknownConnectionError
   */
  this.originalError = err
  this.name = 'UnknownConnectionError'
}

UnknownConnectionError.prototype = Object.create(HabiticaApiError.prototype)

module.exports = {
  HabiticaApiError: HabiticaApiError,
  UnknownConnectionError: UnknownConnectionError
}
</code></pre>
        </article>
    </section>




</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.4.3</a> on Fri Jan 20 2017 17:22:38 GMT-0600 (CST) using the Minami theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>
</body>
</html>
